### ISSUE TITLE
Realm JS migrate from object server 1 to object server 2

### ISSUE THREAD

**Issue by jasonmerino** [realm/realm-js#1705]
## Goals

I am trying to do the upgrade from object server 1 to object server 2. In my testing I have run the commands under the [Migrating from Realm Object Server 1.x](https://realm.io/docs/realm-object-server/latest/#upgrading) heading to upgrade the object server and then I have made changes in my application's code to according to the [Migrating from Realm Object Server 1.x to 2.x section](https://realm.io/docs/javascript/latest/#migrating-from-realm-object-server-1x-to-2x) which suggests I do something like this.

```js
Realm.open(config)
  .then(realm => {
    // you have probably already migrated
  })
  .catch(e => {
    if (e.name == "IncompatibleSyncedRealmError") {
      const backupRealm = new Realm(e.configuration);
      // copy objects from backupRealm
      return;
    }
  });
```

I am seeing that the `IncompatibleSyncedRealmError` is detected as expected and that the data from the object server 1/app gets created as `backupRealm`. That is all good.

But what I am supposed to do to "copy objects from backupRealm" as noted in the comment of the code snippet? Do I copy them into a new blank synced realm?

Above the code snippet on this page it says:

> When a synchronized Realm requiring migration is opened, the Realm file will be copied to a backup location and then deleted so that it can be re-downloaded from the Realm Object Server. An exception will be thrown, and you can choose to migrate your objects from the old Realm

So, since the new object server is a migration from the original, is there a way to access that data and pull it down to the client?

Basically, it seems very unclear to me what I should do after detecting the `IncompatibleSyncedRealmError` exception.

Any help on this would be much appreciated. Thanks!

FWIW I [posted this on StackOverflow](https://stackoverflow.com/questions/49160901/realm-js-migrate-from-object-server-1-to-object-server-2) last week and haven't seen any movement on it, so I'm having to take it to the source.

## Expected Results

NA

## Actual Results

NA

## Steps to Reproduce

Explained above

## Code Sample

Above

## Version of Realm and Tooling

- Realm JS SDK Version: 1.13.0 & 2.2.0
- Node or React Native: RN 0.43.3
- Client OS & Version: ?
- Which debugger for React Native: None

**Comment by ianpward** [comment]
> But what I am supposed to do to "copy objects from backupRealm" as noted in the comment of the code snippet? Do I copy them into a new blank synced realm?

The new synced Realm will download all the data from the server based on the migration done on the backend ROS from 1.x to 2.x - We suggest to copy objects from backupRealm to this new 2.x synced realm if there are additional changes or objects written to the old 1.x Realm before triggering an IncompatibleSyncedRealmError (typically this happens when a user is offline). You will need to open up the backup Realm and compare objects to the new Synced Realm to see if there is any difference and then write the diff to the Synced Realm. I would suggest looking up by PrimaryKeys

**Comment by jasonmerino** [comment]
@ianpward when you say "The new synced Realm will download all the data from the server based on the migration" how is that downloaded. Whenever I have upgraded the app to realm-js v2 it throws the `IncompatibleSyncedRealmError` and so I cannot actually get a reference to the realm you mentioned that should be downloaded from the server. How would I go about doing that?

Another question. Once we upgrade the ROS from 1.x to 2.x we will still have apps out in the wild which will be running realm-js v1 and which will not have updated to the version of the app that runs realm-js v2. I have been running some tests to see what kind of errors we get from the realm-js v1 library when it is pointed at a converted ROS v2 I get this error but I can't seem to find a way to catch it. 

![image](https://user-images.githubusercontent.com/1660279/37540937-0ee67b52-2916-11e8-851b-f18b2dfde24c.png)

How should I go about showing users who are using the old version of the app which uses realm-js v1 a message to tell them to go to the app store and upgrade? I'm not seeing a way to stop the app from crashing once the realm-js v1 is pointed to a ROS v2. 

I guess I could create a feature switch which we flip over once we publish the new app, but it seems like realm-js would expose an error I could key off of to detect if they needed to upgrade or not.

Thanks!

**Comment by ianpward** [comment]
> Whenever I have upgraded the app to realm-js v2 it throws the IncompatibleSyncedRealmError and so I cannot actually get a reference to the realm you mentioned that should be downloaded from the server. How would I go about doing that?

You need to call the Realm.open function again.

> How should I go about showing users who are using the old version of the app which uses realm-js v1 a message to tell them to go to the app store and upgrade? I'm not seeing a way to stop the app from crashing once the realm-js v1 is pointed to a ROS v2.

Correct you should have some sort of view that instructs the user that they need to upgrade in order to proceed.

**Comment by bmunkholm** [comment]
Outdated.
